src/aes/mod.rs

```diff
@@ -1 +1,116 @@
 pub mod registers;
+
+use crate::error::{ErrorKind, Result};
+use crate::memory_map::mmio;
+use registers::Registers;
+
+pub struct AES {
+    registers: Registers,
+}
+
+pub enum Key<'a> {
+    Bits128(&'a [u8; 16]),
+    Bits192(&'a [u8; 24]),
+    Bits256(&'a [u8; 32]),
+}
+
+const ENCRYPT_EXTERNAL_KEY: u8 = 0b_00;
+const DECRYPT_EXTERNAL_KEY: u8 = 0b_01;
+// FIXME User Guide is vague and local generation doesn't make sense
+//const DECRYPT_LOCAL_KEY: u8 = 0b_10;
+
+impl AES {
+    pub fn init() -> Self {
+        Self {
+            registers: Registers::new(mmio::AES),
+        }
+    }
+
+    pub fn encrypt(&mut self, data: &mut [u8], key_size: Key) -> Result<()> {
+        self.external_key(data, key_size, false)
+    }
+
+    pub fn decrypt(&mut self, data: &mut [u8], key_size: Key) -> Result<()> {
+        self.external_key(data, key_size, true)
+    }
+
+    fn external_key(&mut self, data: &mut [u8], key_size: Key, decrypt: bool) -> Result<()> {
```

Comment:
Since it takes a considerable amount of time to set and write a new key to AES, we should instead set the key with some other function and then decrypt/encrypt using previous set key.